pkg/kube/client/cached.go

```diff
@@ -0,0 +1,256 @@
+// Copyright Istio Authors
```

Comment:
For the first glance, it seems like kube controller-runtime, which does implement a cached client for all kinds of resources

Reply:
sigs.k8s.io/controller-runtime/pkg/client/interfaces.go

Reply:
Yes, it is similar. However, controller-runtime has a (very large) number of issues that make it unsuitable for usage in Istio (I know we use it in operator, but that has a much lower bar and different requirements). https://github.com/istio/istio/blob/bdc51cb8e13c94c545d73cd406d2914d6c301d06/pilot/pkg/config/kube/gateway/deploymentcontroller.go#L60-L66 describes a few.

So this is inspired a bit by that library, but taking Istio needs into account. for example, we embedd our StripUnusedFields and SetWatchErrorHandler logic here

Reply:
Yes, that lib is too abstract, and does not consider perf at all. For every resource that is fetched it start up a informer, many many side-effects. WIll have a careful look with the new one here today

Reply:
SGTM. The approach here will have no performance impact, it's just a small wrapper around what we already did for consistency + ease of use